### DISCUSSION TITLE
How should I handle dynamic and static data on a slug route

### DISCUSSION THREAD

**Discussion by JamesSingleton** [vercel/next.js#33439] [Help] [answered]
I have a website that uses sanity.io for the CMS that I render content on a `/[slug].tsx` route. Some of this content can be updated multiple times a day, whereas some of the content is only updated at publish.

I currently have the following in `[slug].tsx`:

```tsx
export const getStaticProps: GetStaticProps = async ({ params }) => {
  const { post, morePosts } = await getClient().fetch(postQuery, {
    slug: params?.slug,
  })

  if (!post) {
    return {
      notFound: true,
    }
  }

  return {
    props: {
      post,
      morePosts: overlayDrafts(morePosts),
    },
  }
}

export const getStaticPaths: GetStaticPaths = async () => {
  const paths = await sanityClient.fetch(postSlugsQuery)

  return {
    paths: paths.map((slug: string) => ({ params: { slug } })),
    fallback: 'blocking',
  }
}
```

I used to have a revalidate key but that meant EVERY single page would get revalidated and that is not good (in my opinion). My current work around is redeploying my app 5+ times a day for new content to be updated which is kind of annoying. Would it be better to just use `getServerSideProps` for something like this?

**Comment by joggienl** [comment, marked as answer]
From the docs:

> Next.js allows you to create or update static pages after you’ve built your site. Incremental Static Regeneration (ISR) enables you to use static-generation on a per-page basis, without needing to rebuild the entire site. With ISR, you can retain the benefits of static while scaling to millions of pages.

So technically if you would add `revalidate: 60` to the return of `getStaticProps`, that would be valid for that particular page (after 60 seconds). So, a single page will get invalidated after that time. It will only be rebuild _after_ you hit that page again. As long as you don't hit the url, nothing will be rebuild automatically.

I noticed the fetch in `getStaticPaths`. Perhaps you are fetching all blogs there, causing them to have more or like the same invalidation time, perhaps that is what you are noticing.

More practically you could generate only the most popular blogs (or the last, or whatever other preference) and generate those. Due to the `blocking` fallback, the rest will be generated once someone hits the url.

Also, what was your `revalidate` time? you could set it to `60 * 60 * 5` to have it invalid once per 5 hours.

**Reply by joggienl** [reply]
Some links:

- https://nextjs.org/docs/api-reference/data-fetching/get-static-props#revalidate
- https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration

**Reply by JamesSingleton** [reply]
I had it set to `revalidate: 7200, // Revalidate every 2 hours` initially. As for the `fetch` call in the `getStaticPaths`, you have to do that when using something like `[slug].tsx` https://nextjs.org/docs/api-reference/data-fetching/get-static-paths#getstaticpaths-return-values.

**Reply by JamesSingleton** [reply]
I don't want to set it for too long as once I update it, it gets posted to social media. So I wouldn't want someone going to it and looking for the update and not be able to find it because it hasn't been updated due to it being set to every 5 hours.

**Reply by joggienl** [reply]
You can set it as short or as long as you please indeed. Shorter means updates get in quicker. Perhaps for your use case something between 5 to 60 minutes would be “better”?

About the `getStaticPaths`, it is indeed mandatory when using a dynamic path for `getStaticProps`. The point I was trying to make there is that it is up to you which paths you return. You don’t have to return all of them. You can also return an empty array to make every first visit blocking.

You could use SSR (`getServerSideProps`) to always get the most up to date content. But in most cases for a blog that is overkill. In that case you could consider the cache control header (with `stale-while-revalidate`) but it would essentially be ISR that way in this situation.

**Reply by JamesSingleton** [reply]
Yeah I have a Sports news website, redshirtsports.xyz which I guess you could classify as a blog for now. So if I had it set to lets say 1 minute, it would only revalidate once someone hits it again. In theory I would just publish my post, post to social media and as long as it is after a minute from publishing it _should_ be updated?

**Reply by joggienl** [reply]
The first one after the timer ends will get the stale content, the regeneration will be triggered and once it is done it will be served.

If you would use 60 as value, perhaps could also (instruct to) visit the page after updating to verify the new content and then post to social media? That would technically take you 60 seconds extra plus generation time (which is very fast in general) 😅.

**Reply by joggienl** [reply]
Technically it is more correct to say that if you hit an url you get the content, if the revalidation time has ended it is considered stale and regeneration will be triggered and if done will be send out as the new version. There are some awesome graphs out there on how it technically works.
